flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/ttl/TtlListState.java

```diff
@@ -0,0 +1,207 @@
+            for (E element : iterable) {
+                list.add(element);
+            }
+            return list;
+        }
+    }
+
+    private List<TtlValue<T>> withTs(List<T> values) {
+        long currentTimestamp = timeProvider.currentTimestamp();
+        List<TtlValue<T>> withTs = new ArrayList<>(values.size());
+        for (T value : values) {
+            Preconditions.checkNotNull(value, "You cannot have null element in a ListState.");
+            withTs.add(TtlUtils.wrapWithTs(value, currentTimestamp));
+        }
+        return withTs;
+    }
+
+    private class IteratorWithCleanup implements Iterator<T> {
+        private final Iterator<TtlValue<T>> originalIterator;
+        private boolean anyUnexpired = false;
+        private boolean uncleared = true;
+        private T nextUnexpired = null;
+
+        private IteratorWithCleanup(Iterator<TtlValue<T>> ttlIterator) {
+            this.originalIterator = ttlIterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            findNextUnexpired();
+            cleanupIfEmpty();
+            return nextUnexpired != null;
+        }
+
+        private void cleanupIfEmpty() {
+            boolean endOfIter = !originalIterator.hasNext() && nextUnexpired == null;
+            if (uncleared && !anyUnexpired && endOfIter) {
+                original.clear();
+                uncleared = false;
+            }
+        }
+
+        @Override
+        public T next() {
+            if (hasNext()) {
+                T result = nextUnexpired;
+                nextUnexpired = null;
+                return result;
+            }
+            throw new NoSuchElementException();
+        }
+
+        // Once a null element is encountered, the subsequent elements will no longer be returned.
+        private void findNextUnexpired() {
+            while (nextUnexpired == null && originalIterator.hasNext()) {
+                TtlValue<T> ttlValue = originalIterator.next();
+                if (ttlValue == null) {
+                    break;
+                }
+                boolean unexpired = !expired(ttlValue);
+                if (unexpired) {
+                    anyUnexpired = true;
+                }
+                if (unexpired || returnExpired) {
+                    nextUnexpired = ttlValue.getUserValue();
+                }
+            }
+        }
+    }
+
+    private class AsyncIteratorWrapper implements StateIterator<T> {
+
+        private final StateIterator<TtlValue<T>> originalIterator;
+
+        public AsyncIteratorWrapper(StateIterator<TtlValue<T>> originalIterator) {
+            this.originalIterator = originalIterator;
+        }
+
+        @Override
+        public <U> StateFuture<Collection<U>> onNext(
+                Function<T, StateFuture<? extends U>> iterating) {
+            Function<TtlValue<T>, StateFuture<? extends U>> ttlIterating =
+                    (item) -> iterating.apply(getElementWithTtlCheck(item));
```

Comment:
The `getElementWithTtlCheck` may return `null`, the user-defined `iterating` function should not be invoked in this case?

Reply:
Good catch, also update the same logic in `TtlMapState`.